### PR TITLE
Schedule DaemonSet Pods in scheduler.

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
part of #59194

**Special notes for your reviewer**:

**Release note**:

```release-note
`ScheduleDaemonSetPods` is an alpha feature (since v1.11) that causes DaemonSet Pods
to be scheduler by default scheduler, instead of Daemonset controller. When it is enabled,
the `NodeAffinity` term (instead of `.spec.nodeName`) is added to the DaemonSet Pods;
this enables the default scheduler to bind the Pod to the target host. If node affinity
of DaemonSet Pod already exists, it will be replaced.

DaemonSet controller will only perform these operations when creating DaemonSet Pods;
and those operations will only modify the Pods of DaemonSet, no changes are made to the
`.spec.template` of DaemonSet.
```

### CODE REVIEW

File: pkg/controller/daemon/util/daemonset_util_test.go [hunk 1]
```diff
@@ -24,8 +24,11 @@ import (
...
 	extensions "k8s.io/api/extensions/v1beta1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/kubernetes/pkg/api/testapi"
+	"k8s.io/kubernetes/pkg/features"
 	kubeletapis "k8s.io/kubernetes/pkg/kubelet/apis"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm"
 )
 
 func newPod(podName string, nodeName string, label map[string]string) *v1.Pod {
```
Comment (wgliang) [inline pkg/controller/daemon/util/daemonset_util_test.go]: `ReplaceDaemonSetPodHostnameNodeAffinity` maybe `ReplaceDaemonSetPodNodeNameNodeAffinity`?

File: pkg/controller/daemon/daemon_controller.go [hunk 2]
```diff
@@ -774,9 +774,14 @@ func (dsc *DaemonSetsController) getNodesToDaemonPods(ds *apps.DaemonSet) (map[s
 	// Group Pods by Node name.
 	nodeToDaemonPods := make(map[string][]*v1.Pod)
 	for _, pod := range claimedPods {
-		nodeName := pod.Spec.NodeName
+		nodeName, err := util.GetTargetNodeName(pod)
```
Comment (aveshagarwal) [inline pkg/controller/daemon/daemon_controller.go]: Here it seems to be changing the behavior if the feature is disabled by returning error if pod.Spec.NodeName is 0.

File: pkg/controller/daemon/util/daemonset_util.go [hunk 3]
```diff
@@ -225,3 +227,43 @@ func AppendNoScheduleTolerationIfNotExist(tolerations []v1.Toleration) []v1.Tole
...
+							algorithm.NodeFieldSelectorKeyNodeName, pod.Namespace, pod.Name)
+					}
+
+					return exp.Values[0], nil
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("no node name found for pod %s/%s", pod.Namespace, pod.Name)
```
Comment (aveshagarwal) [inline pkg/controller/daemon/util/daemonset_util.go]: seems like here we are changing the behavior if the feature is disabled by returning this error which was not checked before.

**Comment by k82cn** [comment]
@janetkuo , PTAL :)

File: pkg/controller/daemon/daemon_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/controller/daemon/daemon_controller.go]: a daemonset pod may never get scheduled (if the target node lacks resources, or is tainted, or not ready, etc). even in cases where the pod eventually gets scheduled, between creation of the unscheduled pod and the binding to the node, the DS controller can easily create lots of extra pods targeted at the node if this doesn't accurately return the ones it already created

File: pkg/controller/daemon/util/daemonset_util.go [hunk 4]
```diff
@@ -225,3 +227,43 @@ func AppendNoScheduleTolerationIfNotExist(tolerations []v1.Toleration) []v1.Tole
 
 	return tolerations
 }
+
+// GetTargetNodeName get the target node name of DaemonSet pods. If ScheduleDaemonSetPods enabled, it will also
+// retrieve node name of pending pods from NodeAffinity; otherwise, `.spec.NodeName` will be return.
```
Comment (janetkuo) [inline pkg/controller/daemon/util/daemonset_util.go]: The comment isn't quite correct. Non-nil `.spec.NodeName` will be returned. Otherwise, we retrieve it from NodeAffinity if ScheduleDaemonSetPods enabled -- which means that as long as `.spec.NodeName` is set we don't check if the feature is enabled or not.

File: pkg/controller/daemon/util/daemonset_util.go [hunk 5]
```diff
@@ -225,3 +227,43 @@ func AppendNoScheduleTolerationIfNotExist(tolerations []v1.Toleration) []v1.Tole
...
+		if len(terms) < 1 {
+			return "", fmt.Errorf("no nodeSelectorTerms in requiredDuringSchedulingIgnoredDuringExecution of pod %s/%s",
+				pod.Namespace, pod.Name)
+		}
+
+		for _, term := range terms {
+			for _, exp := range term.MatchFields {
+				if exp.Key == algorithm.NodeFieldSelectorKeyNodeName &&
+					exp.Operator == v1.NodeSelectorOpIn {
+					if len(exp.Values) != 1 {
```
Comment (janetkuo) [inline pkg/controller/daemon/util/daemonset_util.go]: What if more than one NodeName `MatchFields` is specified? 

Another question: can the pod be scheduled if more than one `exp.Values` is specified? If so, such DaemonSet pod could be scheduled or return an error here.

File: pkg/controller/daemon/util/daemonset_util.go (hunk 3 above)
Comment (janetkuo) [inline pkg/controller/daemon/util/daemonset_util.go]: If the `ScheduleDaemonSetPods` feature is enabled and then disabled, the DaemonSet could have created some pods that don't have `NodeName` specified yet and hit this error. Because this is the first step of `manage`, the DaemonSet will continue being blocked until those pods get scheduled or removed.

File: pkg/controller/daemon/daemon_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (janetkuo) [reply pkg/controller/daemon/daemon_controller.go]: Thanks, that makes sense.

File: pkg/controller/daemon/util/daemonset_util.go (hunk 5 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: Currently, apiserver validation makes sure only one value in `MatchFields` (by `ValidateNodeFieldSelectorRequirement`). For now, `MatchFields` is only used by this feature; I'd like to extend it in coming release.

File: pkg/controller/daemon/util/daemonset_util.go (hunk 3 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: Good point :)

Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: Updated to get `.spec.nodeName` firstly, then `nodeAffinity`.

File: pkg/controller/daemon/daemon_controller.go [hunk 6]
```diff
@@ -774,9 +774,14 @@ func (dsc *DaemonSetsController) getNodesToDaemonPods(ds *apps.DaemonSet) (map[s
...
+		if err != nil {
+			return nil, err
+		}
+
 		nodeToDaemonPods[nodeName] = append(nodeToDaemonPods[nodeName], pod)
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: Lots of the comments that mentioned `nodeToDaemonPods` references it as "a map from nodes to daemon pods (corresponding to ds) running on the nodes". This is no longer true because target node can be a node that doesn't have the pod running yet. 

Please update all those comments, including `getNodesToDaemonPods`, `manage`, etc. 

We probably need to rename some variables too to avoid confusion, such as `daemonPodsRunning`.

File: pkg/controller/daemon/daemon_controller.go [hunk 7]
```diff
@@ -774,9 +774,14 @@ func (dsc *DaemonSetsController) getNodesToDaemonPods(ds *apps.DaemonSet) (map[s
...
 	}
+
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: Because `nodeToDaemonPods` includes pods that's not scheduled yet (no nodeName set), in `podsShouldBeOnNode`, the pods without nodeName set should be killed first when excess daemon pods are found.

File: pkg/controller/daemon/util/daemonset_util.go [hunk 8]
```diff
@@ -177,22 +176,25 @@ func ReplaceDaemonSetPodHostnameNodeAffinity(affinity *v1.Affinity, nodename str
...
+				if exp.Key != algorithm.NodeFieldSelectorKeyNodeName {
+					exps = append(exps, exp)
+				}
 			}
+			term.MatchFields = exps
 		}
 
-		if len(exps) > 0 {
-			term.MatchExpressions = exps
+		if len(term.MatchFields) != 0 || len(term.MatchExpressions) != 0 {
```
Comment (janetkuo) [inline pkg/controller/daemon/util/daemonset_util.go]: This could break if we add another field to `NodeSelectorTerm` in the future. That new field could removed here if the other two fields are both empty.

File: pkg/controller/daemon/daemon_controller.go [hunk 9]
```diff
@@ -774,9 +774,14 @@ func (dsc *DaemonSetsController) getNodesToDaemonPods(ds *apps.DaemonSet) (map[s
 	// Group Pods by Node name.
 	nodeToDaemonPods := make(map[string][]*v1.Pod)
 	for _, pod := range claimedPods {
-		nodeName := pod.Spec.NodeName
+		nodeName, err := util.GetTargetNodeName(pod)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: Added log and continue the loop :)

File: pkg/controller/daemon/util/daemonset_util.go (hunk 3 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: Updated to get `.spec.nodeName` firstly, then `nodeAffinity`.

File: pkg/controller/daemon/daemon_controller.go (hunk 7 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: done, sort the pods by phase & creation timestamp.

File: pkg/controller/daemon/util/daemonset_util.go (hunk 8 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: Yes, that's true; it seems we did not whether other fields in term, otherwise, we leave a empty term here to cover that.

File: pkg/controller/daemon/daemon_controller.go (hunk 6 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: Updated comments to `created for nodes`.

Also check DaemonSet's status which is also confused, e.g. `NumberMisscheduled` should be `NumberMiscreated`. We need to update those description when upgrading to beta.

**Review by bsalamat — COMMENTED** [review]
If I understand correctly, `nodeShouldRunDaemonPod` continues to return `shouldSchedule` after this PR, and `podsShouldBeOnNode` continues to enqueue DS pods for creation at a later time when `shouldSchedule` is false.
This logic does not look correct to me when DS pods are supposed to be scheduled by default scheduler. I think DS controller should create pods as soon as `wantToRun` returned by `nodeShouldRunDaemonPod` is true. DS controller should not even check the predicates anymore. When DS controller creates pods right away, scheduler may preempt other pods to create room for DS pods when the cluster is under resource pressure.

**Comment by k82cn** [comment]
>  I think DS controller should create pods as soon as `wantToRun` returned by `nodeShouldRunDaemonPod` is true.

If feature enabled, this PR only return node selector related errors which makes `wantToRun` equal to `shouldSchedule` and `shouldContinueRunning`: all `true` or `false`. 

>  DS controller should not even check the predicates anymore.

DS still need node match predicates which defines the nodes of DS. The other predicates are not necessary. And we leave the error checking in `nodeShouldRunDaemonPod` to support feature disable. Leaving a TODO here: https://github.com/kubernetes/kubernetes/pull/63223/files#diff-3af9d435092399b4bf3ef61e0387eb36R1317

File: test/integration/daemonset/daemonset_test.go [hunk 10]
```diff
@@ -310,112 +436,178 @@ func forEachStrategy(t *testing.T, tf func(t *testing.T, strategy *apps.DaemonSe
...
+			podInformer := informers.Core().V1().Pods().Informer()
+
+			stopCh := make(chan struct{})
+			informers.Start(stopCh)
+			go dc.Run(5, stopCh)
+
+			// Start Scheduler
+			setupScheduler(t, clientset, informers, stopCh)
+
+			defer close(stopCh)
```
Comment (bsalamat) [inline test/integration/daemonset/daemonset_test.go]: nit: please move this to line 452

File: test/integration/daemonset/daemonset_test.go [hunk 11]
```diff
@@ -310,112 +436,178 @@ func forEachStrategy(t *testing.T, tf func(t *testing.T, strategy *apps.DaemonSe
...
```
Comment (bsalamat) [inline test/integration/daemonset/daemonset_test.go]: ditto: please move it next to the line where the channel is created

File: test/integration/daemonset/daemonset_test.go [hunk 12]
```diff
@@ -310,112 +436,178 @@ func forEachStrategy(t *testing.T, tf func(t *testing.T, strategy *apps.DaemonSe
...
```
Comment (bsalamat) [inline test/integration/daemonset/daemonset_test.go]: ditto

File: test/integration/daemonset/daemonset_test.go [hunk 13]
```diff
@@ -424,20 +616,23 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
...
-		eventClient := clientset.CoreV1().Events(ns.Namespace)
-
 		stopCh := make(chan struct{})
 		informers.Start(stopCh)
 		go dc.Run(5, stopCh)
+
+		// Start Scheduler
+		setupScheduler(t, clientset, informers, stopCh)
+
 		defer close(stopCh)
```
Comment (bsalamat) [inline test/integration/daemonset/daemonset_test.go]: ditto

**Review by bsalamat — COMMENTED** [review]
Thanks, @k82cn for describing the logic. Unfortunately, the current code that supports both the old and new behavior with a feature gate has become confusing. Hopefully we will no longer need to keep the old logic and can simplify the code soon.

File: pkg/controller/daemon/util/daemonset_util.go (hunk 8 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: > Why don't we append the term to nodeSelectorTerms unconditionally?

As `nil or empty` requirements select no objects. 


Did an update to append target node affinity to all terms: it's ORed between terms, and ANDed with a term; and there's no conflict as this target is subset of the node selector.

File: pkg/controller/daemon/daemon_controller.go [hunk 14]
```diff
@@ -1466,14 +1473,25 @@ func (o byCreationTimestamp) Less(i, j int) bool {
...
-func (o podByCreationTimestamp) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
+func (o podByCreationTimestampAndPhase) Len() int      { return len(o) }
+func (o podByCreationTimestampAndPhase) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
 
-func (o podByCreationTimestamp) Less(i, j int) bool {
-	if o[i].CreationTimestamp.Equal(&o[j].CreationTimestamp) {
-		return o[i].Name < o[j].Name
+func (o podByCreationTimestampAndPhase) Less(i, j int) bool {
+	// Unscheduled Pod first.
+	if len(o[i].Spec.NodeName) == 0 && len(o[j].Spec.NodeName) != 0 {
```
[the remark this answers is not in the document]
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: I removed the phase (`.spec.NodeName`) condition in sort; as the newest one is the 'correct' one we created: controller should delete the oldest one, and the check status according to the newest one (also give an update to check the status based on newest one).

@janetkuo

[the remark this answers is not in the document]
Comment (bsalamat) [reply pkg/controller/daemon/daemon_controller.go]: But the original code used to return the oldest one first. So, I think we should keep the same ordering.

File: pkg/controller/daemon/daemon_controller.go [hunk 15]
```diff
@@ -1472,8 +1482,11 @@ func (o podByCreationTimestamp) Len() int      { return len(o) }
 func (o podByCreationTimestamp) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
 
 func (o podByCreationTimestamp) Less(i, j int) bool {
-	if o[i].CreationTimestamp.Equal(&o[j].CreationTimestamp) {
-		return o[i].Name < o[j].Name
+	// Early CreationTimestamp first
```
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: Since you are not changing the logic here, please revert this function to what it was before to minimize debugging surface in case any issue is found later.

Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: Done, reverted it :)

File: pkg/controller/daemon/daemon_controller.go [hunk 16]
```diff
@@ -1099,7 +1108,8 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *apps.DaemonSet, hash
 				// Sort the daemon pods by creation time, so that the oldest is first.
 				daemonPods, _ := nodeToDaemonPods[node.Name]
 				sort.Sort(podByCreationTimestamp(daemonPods))
-				pod := daemonPods[0]
+				// Check the newest one; the oldest pods will be deleted later.
+				pod := daemonPods[len(daemonPods)-1]
```
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: Is this a new change or I missed it before? Why is this change needed?

File: pkg/controller/daemon/daemon_controller.go (hunk 7 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: Just revert it; the node without `nodeName` maybe the new one we want to create.

File: pkg/controller/daemon/daemon_controller.go (hunk 16 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: A new change: we should check the status according the new pods, the old one will be deleted. Anyway, that's not related to this PR; revert it, and I'll open another PRs for this.

**Review by bsalamat — COMMENTED** [review]
/lgtm

File: test/integration/daemonset/daemonset_test.go [hunk 17]
```diff
@@ -310,112 +436,179 @@ func forEachStrategy(t *testing.T, tf func(t *testing.T, strategy *apps.DaemonSe
...
-		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
-		validateDaemonSetStatus(dsClient, ds.Name, 1, t)
+		validateFailedPlacementEvent(eventClient, t)
 	})
 }
 
-func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
+// TestInsufficientCapacityNodeDaemonSetCreateButNotLaunchPod tests that the DaemonSet will
+// create Pods for the node with insufficient resource, but kube-scheduler did not launch it
+// when ScheduleDaemonSetPods is enabled.
```
Comment (janetkuo) [inline test/integration/daemonset/daemonset_test.go]:
```go
// TestInsufficientCapacityNodeDaemonSetCreateButNotLaunchPod tests that when "ScheduleDaemonSetPods" 
// feature is enabled, the DaemonSet should create Pods for all the nodes regardless of available resource
// on the nodes, and kube-scheduler should not schedule Pods onto the nodes with insufficient resource.
```

File: test/integration/daemonset/daemonset_test.go [hunk 18]
```diff
@@ -447,6 +643,17 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
 			t.Fatalf("Failed to create node: %v", err)
 		}
```
Comment (janetkuo) [inline test/integration/daemonset/daemonset_test.go]: Create another node here with enough resource to schedule, and the scheduler should schedule a pod on it.

File: test/integration/daemonset/daemonset_test.go [hunk 19]
```diff
@@ -310,112 +436,179 @@ func forEachStrategy(t *testing.T, tf func(t *testing.T, strategy *apps.DaemonSe
...
-		validateDaemonSetStatus(dsClient, ds.Name, 1, t)
+		validateFailedPlacementEvent(eventClient, t)
 	})
 }
 
-func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
+// TestInsufficientCapacityNodeDaemonSetCreateButNotLaunchPod tests that the DaemonSet will
+// create Pods for the node with insufficient resource, but kube-scheduler did not launch it
+// when ScheduleDaemonSetPods is enabled.
+func TestInsufficientCapacityNodeWhenScheduleDaemonSetPodsEnabled(t *testing.T) {
```
Comment (janetkuo) [inline test/integration/daemonset/daemonset_test.go]: Enabling this feature will break some DaemonSet unit tests, such as `TestInsufficientCapacityNodeDaemonDoesNotLaunchPod`

File: pkg/controller/daemon/daemon_controller.go (hunk 7 above)
Comment (janetkuo) [reply pkg/controller/daemon/daemon_controller.go]: > Just revert it; the node without `nodeName` maybe the new one we want to create.

I'm not sure if I understand why this is reverted. If a DaemonSet pod is already scheduled on this node, shouldn't other pending pods be deleted first?

File: pkg/controller/daemon/util/daemonset_util.go [hunk 20]
```diff
@@ -177,23 +176,27 @@ func ReplaceDaemonSetPodHostnameNodeAffinity(affinity *v1.Affinity, nodename str
...
 		}
 
-		if len(exps) > 0 {
-			term.MatchExpressions = exps
-			nodeSelectorTerms = append(nodeSelectorTerms, term)
-		}
-	}
+		// It's ORed between terms, it's required to append target node selector to each term.
+		// DaemonSet controller already makes sure this node is one of target which is subset of
+		// terms, so no conflict.
```
[the remark this answers is not in the document]
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: For example, if a Pod has `term_1, term_2, and term_3`, the host in this func matches **at least** one term (as the term are ORed, and the host passed node selector predicates); we can not just append a new `term_4`. In such case, maybe we can just replace NodeAffinity with `MatchFields` term.

File: pkg/controller/daemon/util/daemonset_util_test.go [hunk 21]
```diff
@@ -333,10 +334,117 @@ func TestReplaceDaemonSetPodHostnameNodeAffinity(t *testing.T) {
 		},
 	}
```
[the remark this answers is not in the document]
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util_test.go]: done

File: test/integration/daemonset/daemonset_test.go (hunk 18 above)
Comment (k82cn) [reply test/integration/daemonset/daemonset_test.go]: done.

File: pkg/controller/daemon/daemon_controller.go [hunk 22]
```diff
@@ -962,9 +969,11 @@ func (dsc *DaemonSetsController) syncNodes(ds *apps.DaemonSet, podsToDelete, nod
 
 				podTemplate := &template
 
-				if false /*disabled for 1.10*/ && utilfeature.DefaultFeatureGate.Enabled(features.ScheduleDaemonSetPods) {
+				if utilfeature.DefaultFeatureGate.Enabled(features.ScheduleDaemonSetPods) {
```
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: We should enable this feature by default in 1.11.

Comment (janetkuo) [reply pkg/controller/daemon/daemon_controller.go]: We can't enable an alpha feature by default.

Comment (bsalamat) [reply pkg/controller/daemon/daemon_controller.go]: This cannot remain disabled in 1.11. Rescheduler is already removed from the code-base. If critical daemonsets cannot be scheduled, preemption must create room for them and DS controller is incapable of performing preemption.

**Comment by bsalamat** [comment]
@k82cn Please add a very clear release note about this feature.

File: pkg/controller/daemon/daemon_controller.go (hunk 22 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: @bsalamat , here's the code about critical pod in daemonset: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L1429 .

Comment (bsalamat) [reply pkg/controller/daemon/daemon_controller.go]: Thanks, Klaus. So, my initial concern is valid. DS controller does not run "resource check" for critical pods. This means that it creates critical DS Pods regardless of the resources available on the nodes and it relies on "Rescheduler" to free up resources on the nodes if necessary. In the absence of Rescheduler, it is important to let default scheduler schedule DS Pods. Otherwise, critical DS pods may never be scheduled when their corresponding nodes are out of resources.

Comment (bsalamat) [reply pkg/controller/daemon/daemon_controller.go]: I arranged with @ravisantoshgudimetla to make Rescheduler aware of Pod priority and add it back to help create room for critical DS Pods. So, this PR can remain as is (no need to enable the feature in 1.11).

**Comment by bsalamat** [comment]
Please change the following phrase in the release notes:
s/that makes DaemonSet Pods are scheduled by default scheduler/that causes DaemonSet Pods to be scheduler by default scheduler/

otherwise, LGTM

File: pkg/controller/daemon/daemon_controller.go (hunk 22 above)
Comment (bsalamat) [reply pkg/controller/daemon/daemon_controller.go]: I have to add that it was @ravisantoshgudimetla's idea to add priority awareness and use Rescheduler in 1.11. It removes a blocker in moving priority and preemption to Beta.

File: pkg/controller/daemon/util/daemonset_util.go [hunk 23]
```diff
@@ -136,19 +135,19 @@ func SplitByAvailablePods(minReadySeconds int32, pods []*v1.Pod) ([]*v1.Pod, []*
 	return availablePods, unavailablePods
 }
 
-// ReplaceDaemonSetPodHostnameNodeAffinity replaces the 'kubernetes.io/hostname' NodeAffinity term with
-// the given "nodeName" in the "affinity" terms.
-func ReplaceDaemonSetPodHostnameNodeAffinity(affinity *v1.Affinity, nodename string) *v1.Affinity {
+// ReplaceDaemonSetPodNodeNameNodeAffinity replaces the NodeAffinity by a new NodeAffinity with
+// the given "nodeName" in the "affinityterms.
```
Comment (janetkuo) [inline pkg/controller/daemon/util/daemonset_util.go]: nit:
```go
// ReplaceDaemonSetPodNodeNameNodeAffinity replaces the RequiredDuringSchedulingIgnoredDuringExecution 
// NodeAffinity of the given affinity with a new NodeAffinity that selects the given nodeName.
// Note that this function assumes that no NodeAffinity conflicts with the selected nodeName.
```

File: pkg/controller/daemon/daemon_controller.go [hunk 24]
```diff
@@ -962,9 +969,11 @@ func (dsc *DaemonSetsController) syncNodes(ds *apps.DaemonSet, podsToDelete, nod
...
 					podTemplate = template.DeepCopy()
-					podTemplate.Spec.Affinity = util.ReplaceDaemonSetPodHostnameNodeAffinity(
+					// The pod's NodeAffinity will be updated to make sure the Pod is bound
+					// to the target node by default scheduler.
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: Add: 
```go
// It is safe to do so because there should be no conflicting node affinity with the target node. 
```

**Review by janetkuo — APPROVED** [review]
Just some nits on comments. LGTM otherwise. Please squash commits. Thanks!

File: pkg/controller/daemon/daemon_controller.go (hunk 24 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: done

File: pkg/controller/daemon/util/daemonset_util.go (hunk 23 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: done

File: pkg/controller/daemon/daemon_controller.go [hunk 25]
```diff
@@ -850,7 +857,7 @@ func (dsc *DaemonSetsController) podsShouldBeOnNode(
 		// If daemon pod is supposed to be running on node, but more than 1 daemon pod is running, delete the excess daemon pods.
 		// Sort the daemon pods by creation time, so the oldest is preserved.
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: nit: scheduled pod is preserved first; if more than one pod can be preserved, the oldest pod is preserved.
